File: Example,compute NFD.R
# @author: J.W.Spaak
# Example how to compute the ND and FD for a given differential equation setting

# The script is originally written in python, we run python code from within R
# make sure that python is installed on the system used
# This is not part of the install.packages("reticulate") command!
library(reticulate)
setwd("C:/Users/Jurg/Dropbox/Doktorat/Projects/P5_Fitness and niche differences/3_Programs/")

# loads the relevant python code
source_python("numerical_NFD.py")

# create the differential equation system
n_spec <- 2 # number of species in the system, must be an integer
set.seed(0) # set random seed for reproduce ability

# Lotka-Volterra model
A <- matrix(runif(n_spec^2,0,1), n_spec, n_spec) # interaction matrix
diag(A) <- runif(n_spec, 1,2) # to ensure coexistence
mu <- runif(n_spec,1,2) # intrinsic growth rate
test_f <- function(N){
  return(mu - A%*%N)
}

# compute relevant parameters with python
# the parameter `from_R = TRUE` changes data types from R to python
pars <- find_NFD(test_f, n_spec, from_R = TRUE)
ND <- pars$ND
NO <- pars$NO
FD <- pars$FD
c <- pars$c

# manualy check results for the two species case
# see appendix for proof of correctness
NO_check = sqrt(A[1,2]*A[2,1]/(A[1,1]*A[2,2]))*c(1,1)
ND_check = 1-NO_check
FD_check = 1- rev(mu)/mu*sqrt(c(A[1,2]*A[1,1]/A[2,1]/A[2,2],
                                            A[2,1]*A[2,2]/A[1,2]/A[1,1]))
c_check = sqrt(c(A[1,2]*A[2,2]/A[2,1]/A[1,1],
                            A[2,1]*A[1,1]/A[1,2]/A[2,2]))

###############################################################################
# Switching to multispecies case
# create the differential equation system
n_spec <- 10 # nuber of species in the system

# Lotka-Volterra model
A <- matrix(runif(n_spec^2,0,1), n_spec, n_spec) # interaction matrix
diag(A) <- runif(n_spec, n_spec, n_spec+1) # to ensure coexistence in the example
mu <- runif(n_spec,1,2) # intrinsic growth rate
test_f <- function(N){
  return(mu - A%*%N)
}

# compute relevant parameters with software
pars <- find_NFD(test_f, n_spec, from_R = TRUE)
ND <- pars$ND
NO <- pars$NO
FD <- pars$FD
c <- pars$c

# manualy check results for the two species case
# see appendix for proof of correctness
NO_check_m <- rep(0, n_spec)
FD_check_m <- rep(0, n_spec)
for (i in 1:n_spec){
  denominator = 0
  numerator = 0
  for (j in 1:n_spec){
    if (i==j) next
    numerator = numerator + pars$N_star[i,j]*A[i,j]
    denominator =denominator + pars$N_star[i,j]*sqrt(A[i,j]/A[j,i]*A[i,i]*A[j,j])
  NO_check_m[i] = numerator/denominator
  FD_check_m[i] = 1-denominator/mu[i]
  }
}

File: Example,compute NFD.py
"""
@author: J.W.Spaak
Example how to compute the ND and FD for a given differential equation setting
"""

import numpy as np
from numerical_NFD import NFD_model

# create the differential equation system
n_spec = 2 # number of species in the system
np.random.seed(6) # set random seed for reproduce ability

# Lotka-Volterra model
A = np.random.uniform(0,1,(n_spec,n_spec)) # interaction matrix
np.fill_diagonal(A,np.random.uniform(1,2,n_spec)) # to ensure coexistence
mu = np.random.uniform(1,2,n_spec) # intrinsic growth rate
def test_f(N):
    return mu - np.dot(A,N)

# compute relevant parameters with software
pars = NFD_model(test_f, n_spec)
ND, NO, FD, c = pars["ND"], pars["NO"], pars["FD"], pars["c"]
# manualy check results for the two species case
# see appendix for proof of correctness
NO_check = np.sqrt(np.array([A[0,1]*A[1,0]/A[1,1]/A[0,0],
                       A[0,1]*A[1,0]/A[1,1]/A[0,0]]))
ND_check = 1-NO
FD_check = 1- mu[::-1]/mu*np.sqrt(np.array([A[0,1]*A[0,0]/A[1,0]/A[1,1],
                       A[1,0]*A[1,1]/A[0,1]/A[0,0]]))
c_check = np.sqrt(np.array([A[0,1]*A[1,1]/A[1,0]/A[0,0],
                      A[1,0]*A[0,0]/A[0,1]/A[1,1]]))

# precision of output
prec = 4
print("Results of two species case:\n")
print("\t Software\t\t Manual check\t\t Rel. Difference\n")
print("ND:\t", np.round(ND,prec), "\t", np.round(ND_check,prec),
      "\t", np.abs(ND-ND_check)/ND_check)
print("NO:\t", np.round(NO,prec), "\t", np.round(NO_check,prec),
      "\t", np.abs(NO-NO_check)/NO_check)
print("FD:\t", np.round(FD,prec), "\t", np.round(FD_check,prec),
      "\t", np.abs(FD-FD_check)/FD_check)
print("c:\t", np.round(c[[0,1],[1,0]],prec), "\t", np.round(c_check,prec),
      "\t", np.abs(c[[0,1],[1,0]]-c_check)/c_check)


###############################################################################
# Switching to multispecies case
# create the differential equation system
n_spec = 10 # nuber of species in the system

# Lotka-Volterra model
A = np.random.uniform(0,1,(n_spec,n_spec)) # interaction matrix

# to ensure coexistence increase diagonal values
np.fill_diagonal(A,np.random.uniform(n_spec,n_spec+1,n_spec)) 
mu = np.random.uniform(1,2,n_spec) # intrinsic growth rate
def test_f(N):
    return mu - np.dot(A,N)

pars = NFD_model(test_f, n_spec)
ND_m, NO_m, FD_m, c_m = pars["ND"], pars["NO"], pars["FD"], pars["c"]

NO_check_m = np.empty(n_spec)
FD_check_m = np.empty(n_spec)
for i in range(n_spec):
    denominator = 0
    numerator = 0
    for j in range(n_spec):
        if i==j:
            continue
        numerator += pars["N_star"][i,j]*A[i,j]
        denominator += pars["N_star"][i,j]*np.sqrt(A[i,j]/A[j,i]*A[i,i]*A[j,j])
    NO_check_m[i] = numerator/denominator
    FD_check_m[i] = 1-denominator/mu[i]
    
# printing layout is optimized for 6 species
def print_function(var, var_check, name):
    rel_diff = np.round(np.abs(var-var_check)/var_check,prec)
    var = np.round(var,prec)
    var_check = np.round(var_check,prec)
    print(name+":\t", var[:2],"\t", var_check[:2], "\t", rel_diff[:2])
    for i in range(1,len(var)//2):
        ind = [2*i,2*i+1]
        print("\t", var[ind],"\t", var_check[ind], "\t", rel_diff[ind])
    if len(var)%2==1:
        print("\t", var[-1],"\t\t", var_check[-1], "\t\t", rel_diff[-1])
    print()
        

print("\n\nResults of multi species case:\n")
print("\t Software\t\t Manual check\t\t Rel. Difference\n")
print_function(NO_m, NO_check_m, "NO")
print_function(FD_m, FD_check_m, "FD")
